api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/operations/XMLTransformation.java

```diff
@@ -0,0 +1,115 @@
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.w3c.dom.Document;
+
+import edu.cornell.mannlib.vitro.webapp.dynapi.components.OperationResult;
+import edu.cornell.mannlib.vitro.webapp.dynapi.components.Parameter;
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.Data;
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.DataStore;
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.SimpleDataView;
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.conversion.InitializationException;
+import edu.cornell.mannlib.vitro.webapp.utils.configuration.Property;
+
+public class XMLTransformation extends AbstractOperation {
+
+    private static final Log log = LogFactory.getLog(XMLTransformation.class);
+    private Parameter xsltParam;
+    private Parameter inputXmlParam;
+    private Parameter outputXmlParam;
+
+    @Property(uri = "https://vivoweb.org/ontology/vitro-dynamic-api#xslt", minOccurs = 1, maxOccurs = 1)
+    public void setXsltParam(Parameter xsltParam) throws InitializationException {
+        this.xsltParam = xsltParam;
+        inputParams.add(xsltParam);
+    }
+
+    @Property(uri = "https://vivoweb.org/ontology/vitro-dynamic-api#inputXml", minOccurs = 1, maxOccurs = 1)
+    public void setInputXmlParam(Parameter inputXmlParam) throws InitializationException {
+        this.inputXmlParam = inputXmlParam;
+        inputParams.add(inputXmlParam);
+    }
+
+    @Property(uri = "https://vivoweb.org/ontology/vitro-dynamic-api#outputXml", minOccurs = 1, maxOccurs = 1)
+    public void setOutputXml(Parameter outputXmlParam) throws InitializationException {
+        this.outputXmlParam = outputXmlParam;
+        outputParams.add(outputXmlParam);
+    }
+
+    @Override
+    public OperationResult runOperation(DataStore dataStore) throws Exception {
+        String is = SimpleDataView.getStringRepresentation(inputXmlParam.getName(), dataStore);
+        String styles = SimpleDataView.getStringRepresentation(xsltParam.getName(), dataStore);
+        ByteArrayOutputStream output = transform(is, styles);
+        Data outputData = new Data(outputXmlParam);
+        outputData.setRawString(output.toString(StandardCharsets.UTF_8));
+        outputData.initializeFromString();
+        dataStore.addData(outputXmlParam.getName(), outputData);
+        return OperationResult.ok();
+    }
+
+    private ByteArrayOutputStream transform(String input, String styles) throws Exception {
+        InputStream inputStream = IOUtils.toInputStream(input, StandardCharsets.UTF_8);
+        InputStream styleInputStream = IOUtils.toInputStream(styles, StandardCharsets.UTF_8);
+        Source stylesource = new StreamSource(styleInputStream);
+        ByteArrayOutputStream output = new ByteArrayOutputStream();
+        TransformerFactory transformerFactory = TransformerFactory.newInstance();
+        Transformer transformer = transformerFactory.newTransformer(stylesource);
```

Comment:
Maybe we could have our own more powerful factory, which might include XSLT transformation, but also RML transformation, and even some our custom transformation implemented in some Java classes.

Reply:
I think we can safely do that later.